File: ast/declaration/declaration.go
package declaration

import (
	"github.com/panda-io/micro-panda/ast/ast"
	"github.com/panda-io/micro-panda/ast/expression"
)

type Attribute struct {
	Position int
	Name     string
	Text     string
	Values   map[string]*expression.Literal
}

type DeclarationBase struct {
	ast.NodeBase
	Attributes []*Attribute
	Public     bool
	Name       *expression.Identifier
	Qualified  string
}

func (b *DeclarationBase) HasAttribute(attribute string) bool {
	for _, a := range b.Attributes {
		if a.Name == attribute {
			return true
		}
	}
	return false
}

func (b *DeclarationBase) GetAttribute(attribute string, name string) *expression.Literal {
	for _, a := range b.Attributes {
		if a.Name == attribute {
			return a.Values[name]
		}
	}
	return nil
}

func (b *DeclarationBase) QualifiedName() string {
	return b.Qualified
}

File: MicroPanda/AST/Declaration/Declaration.cs
namespace MicroPanda.AST.Declaration;

using Node;
using Expression;

internal abstract class Declaration : Node
{
    internal List<Attribute> Attributes = [];
    internal bool IsPublic { get; set; }
    internal Identifier? Name { get; set; }
    internal string? QualifiedName { get; set; }
    internal abstract void ValidateType(Context context);
    internal abstract void Validate(Context context);

    internal bool HasAnnotation(string annotation)
    {
        foreach (var attribute in Attributes)
        {
            if (attribute.Name == annotation)
            {
                return true;
            }
        }
        return false;
    }

    internal bool HasAnnotationField(string annotation, string field)
    {
        foreach (var attribute in Attributes)
        {
            if (attribute.Name == annotation)
            {
                return attribute.Values.ContainsKey(field);
            }
        }
        return false;
    }

    internal Literal? GetAnnotationField(string annotation, string field)
    {
        foreach (var attribute in Attributes)
        {
            if (attribute.Name == annotation)
            {
                return attribute.Values[field];
            }
        }
        return null;
    }
}

// Annotation member types: token.INT, token.FLOAT, token.STRING, token.BOOL
internal class Attribute
{
    internal int Position { get; set; }
    internal string? Name { get; set; }
    internal Dictionary<string, Literal> Values = [];
}